### ISSUE TITLE
HTTP tracker announce fails due to read timeout on slow trackers

### ISSUE THREAD

**Issue by qix67** [arvidn/libtorrent#4567]
**Please provide the following information**

libtorrent version (or branch):
1.2.5

platform/architecture:
linux/x86_64

compiler and compiler version:
gcc 9.2.0

please describe what symptom you see, what you would expect to see instead and
how to reproduce it.

I have encounted a private HTTP tracker reported as "not working" by qBittorrent. The problem occurs during announce request. The client closes the connection after ~5 seconds.

I picked up announce request using tcpdump and then tried to send the same announce request manually using curl. The request works properly. I noticed the tracker immediatly accepts connection but then takes more than 11s to reply.

In src/http_connection.cpp line 239, replacing:
m_read_timeout = seconds(5);
by
m_read_timeout = seconds(20);
makes the tracker available however it is surely not a good solution.

It whould be great to have an option to adjust this timeout.

**Comment by FranciscoPombal** [comment]
You can adjust with the `stop_tracker_timeout` setting
https://github.com/arvidn/libtorrent/blob/a9968916ca82366f1c236af59aaecb9bc94ffe73/include/libtorrent/settings_pack.hpp#L860

**Comment by qix67** [comment]
I have tried this variable but I think the description is misleading because it speaks about "stopped message" and in my case, it is a "start event". Moreover, it is nowhere written this value is divided by 5 (line 246).

**Comment by FranciscoPombal** [comment]
You're right, that is only for `stop` events. Otherwise, this is the timeout you want to configure:
https://github.com/arvidn/libtorrent/blob/a9968916ca82366f1c236af59aaecb9bc94ffe73/include/libtorrent/settings_pack.hpp#L848

Which as a default value of `30`:
https://github.com/arvidn/libtorrent/blob/a9968916ca82366f1c236af59aaecb9bc94ffe73/src/settings_pack.cpp#L218

You can see how the distinction between event types here:
https://github.com/arvidn/libtorrent/blob/a9968916ca82366f1c236af59aaecb9bc94ffe73/src/http_tracker_connection.cpp#L215

See also the related setting `tracker_receive_timeout`.

But yes, I have no idea why this happens:
https://github.com/arvidn/libtorrent/blob/a9968916ca82366f1c236af59aaecb9bc94ffe73/src/http_connection.cpp#L240

Especially since the timer will then be set to `std::min(m_read_timeout, m_completion_timeout)`. So if the user uses a completion timeout of 30, for example, the actual timeout will be 6. Why is this @arvidn?

**Comment by arvidn** [comment]
@FranciscoPombal The idea with the two timeouts are that as long as there is *some* data coming back, we'll give it the full completion-timeout to complete. However, if we don't hear *anything* for read-timeout, we give up sooner.

I'm open to increase timeouts. Is the read timeout not configurable?

**Comment by FranciscoPombal** [comment]
@arvidn 
> Is the read timeout not configurable?

Apparently not.

> @FranciscoPombal The idea with the two timeouts are that as long as there is _some_ data coming back, we'll give it the full completion-timeout to complete. However, if we don't hear _anything_ for read-timeout, we give up sooner.

Assume a `tracker_completion_timeout` of `30` seconds. What if the tracker sends a bit of data, then nothing for 7 seconds, then the rest at the 8th second? With the way things are now, libtorrent would have given up. I don't think this is expected. Is there really a need to be so aggressive with tracker request timeouts? These are connections that are only made every 30 - 45 minutes or so anyway. Perhaps the actual timeout could just be the value of `m_completion_timeout` by default; then there could be a setting called `aggressive_tracker_timer` or something similar that enabled the current "timer is shorter if no data is being read" logic.

**Comment by arvidn** [comment]
>  Is there really a need to be so aggressive with tracker request timeouts?

Probably not. (but it has to be a pretty shitty database to take this long to respond too :) )

**Comment by arvidn** [comment]
does this look reasonable? https://github.com/arvidn/libtorrent/pull/4569

**Comment by arvidn** [comment]
the case where it matters to timeout quickly is where there are multiple trackers waiting to be tried, and the ones that aren't working must fail before moving on to the next.

**Comment by FranciscoPombal** [comment]
@arvidn 
> the case where it matters to timeout quickly is where there are multiple trackers waiting to be tried, and the ones that aren't working must fail before moving on to the next.

Since you're removing the read timer completely, it would probably then make sense to set the default of the completion timer a bit lower, to maybe 10 or less.

**Comment by arvidn** [comment]
patch has landed
